### PR TITLE
🌱 Prep main branch for v1.7 development

### PR DESCRIPTION
**Pull request by cahillsf**
**What this PR does / why we need it**:
This PR prepares the main branch for development.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of:  [#9800](https://github.com/kubernetes-sigs/cluster-api/issues/9800)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -531,7 +531,7 @@ generate-doctoc:
 	TRACE=$(TRACE) ./hack/generate-doctoc.sh
 
 .PHONY: generate-e2e-templates
-generate-e2e-templates: $(KUSTOMIZE) $(addprefix generate-e2e-templates-, v1.0 v1.4 v1.5 main) ## Generate cluster templates for all versions
+generate-e2e-templates: $(KUSTOMIZE) $(addprefix generate-e2e-templates-, v1.0 v1.5 v1.6 main) ## Generate cluster templates for all versions
```
Comment (sbueringer) [inline Makefile:534]: Not sure if this change will work without the corresponding changes to our tests that are using these files 
(please ignore if this PR is just WIP :))

Comment (cahillsf) [reply Makefile:534]: 👋 yep just trying to get ahead 😄 .  will review more thoroughly once we have cut the `v1.6.0` release this week

**Review by cahillsf — COMMENTED** [review]
is this still true? looks like the hardcoded SHA refs were removed from the templates in the last cycle: https://github.com/kubernetes-sigs/cluster-api/pull/9097/files#diff-bad8038a4f09fe4c891e6c9021ec75602c324487810f68cce202b71967ac171aL11-L12

File: test/e2e/clusterctl_upgrade_test.go [hunk 2]
```diff
@@ -62,24 +62,24 @@ var _ = Describe("When testing clusterctl upgrades (v1.4=>current)", func() {
...
-			InitWithBootstrapProviders:      []string{"kubeadm:v1.4.5"},
-			InitWithControlPlaneProviders:   []string{"kubeadm:v1.4.5"},
-			InitWithInfrastructureProviders: []string{"docker:v1.4.5"},
+			// use the latest version for the contract (which is v1.6.X for v1beta1).
+			InitWithCoreProvider:            "cluster-api:v1.5.4",
+			InitWithBootstrapProviders:      []string{"kubeadm:v1.5.4"},
+			InitWithControlPlaneProviders:   []string{"kubeadm:v1.5.4"},
+			InitWithInfrastructureProviders: []string{"docker:v1.5.4"},
 			InitWithProvidersContract:       "v1beta1",
-			// NOTE: If this version is changed here the image and SHA must also be updated in all DockerMachineTemplates in `test/e2e/data/infrastructure-docker/v1.4/bases.
```
Comment (cahillsf) [inline test/e2e/clusterctl_upgrade_test.go:73]: is this still true? looks like the hardcoded SHA refs were removed from the templates in the last cycle: https://github.com/kubernetes-sigs/cluster-api/pull/9097/files#diff-bad8038a4f09fe4c891e6c9021ec75602c324487810f68cce202b71967ac171aL11-L12

File: test/e2e/data/shared/v1.6/metadata.yaml [hunk 3]
```diff
@@ -1,6 +1,15 @@
 apiVersion: clusterctl.cluster.x-k8s.io/v1alpha3
 kind: Metadata
 releaseSeries:
+  - major: 1
+    minor: 7
+    contract: v1beta1
```
Comment (sbueringer) [inline test/e2e/data/shared/v1.6/metadata.yaml:6]:
```suggestion
```

This should not contain v1.7. (the metadata.yaml of the 1.6 release does not contain the 1.7 release)

File: test/e2e/clusterctl_upgrade_test.go (hunk 2 above)
Comment (sbueringer) [reply test/e2e/clusterctl_upgrade_test.go:73]: You're right. We should drop this comment here and in other places below (this only still applies to 1.0)

File: test/e2e/clusterctl_upgrade_test.go [hunk 4]
```diff
@@ -62,24 +62,24 @@ var _ = Describe("When testing clusterctl upgrades (v1.4=>current)", func() {
...
-			InitWithKubernetesVersion: "v1.27.3",
+			// NOTE: If this version is changed here the image and SHA must also be updated in all DockerMachineTemplates in `test/e2e/data/infrastructure-docker/v1.5/bases.
+			InitWithKubernetesVersion: "v1.28.0",
 			WorkloadKubernetesVersion: "v1.27.3",
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade_test.go:75]: We should also bump the WorkloadKubernetesVersion to 1.28.0

(same for the next test case)

**Comment by sbueringer** [comment]
Just a few smaller findings. We should be careful in combination with this PR: #9265. Whatever gets merged second has to be adjusted. 

@adilGhaffarDev If we can get a quick round of fixes in #9265. I would prefer merging #9265 first

File: test/e2e/clusterctl_upgrade_test.go (hunk 4 above)
Comment (sbueringer) [reply test/e2e/clusterctl_upgrade_test.go:75]: We should probably add this comment from the release tasks here inline as a godoc comment:
> Note: Both InitWithKubernetesVersion and WorkloadKubernetesVersion should be the highest mgmt cluster version supported by the respective Cluster API version.

**Comment by adilGhaffarDev** [comment]
> @adilGhaffarDev If we can get a quick round of fixes in #9265. I would prefer merging #9265 first

I will address the comments today on my PR.

File: test/e2e/clusterctl_upgrade_test.go (hunk 4 above)
Comment (cahillsf) [reply test/e2e/clusterctl_upgrade_test.go]: pushed up some changes, i believe they accomplish the suggestions (please let me know if otherwise)

Comment (sbueringer) [reply test/e2e/clusterctl_upgrade_test.go]: Looks perfect. Thank you!

**Comment by sbueringer** [comment]
Thx, looks good so far. Let's wait for merge of the other PR and then after a rebase (and using the release markers) we should be good

**Comment by chrischdi** [comment]
#9265 just merged, so this is ready to get rebased @cahillsf 👍

**Review by chrischdi — COMMENTED** [review]
/lgtm

**Comment by sbueringer** [comment]
/lgtm

I think we don't have to squash for the release notes tool. But looks like we have to because of the `needs-rebase` label.

@cahillsf Can you please squash the commits?

**Comment by sbueringer** [comment]
/approve


/hold
(for squash)

**Comment by chrischdi** [comment]
@cahillsf : I don't want to push pressure, but it would be helpful to get this rebased and squashed so we can merge it. This then would unblock work on k8s v1.29 :-)

**Comment by cahillsf** [comment]
yep np @chrischdi its on my list, US workday coming to a close soon will squash as soon as i have a spare moment

**Comment by cahillsf** [comment]
@chrischdi just squashed and rebased, hopefully the tests will pass. does need another lgtm/approve though 😄

**Comment by sbueringer** [comment]
/lgtm

/hold cancel

Should now merge after tests

**Comment by sbueringer** [comment]
Thx @cahillsf !!

**Pull request merged**